### PR TITLE
adds removal of non active contracts from user

### LINKED ISSUE #104: cleanup UserContractSets after contracts are closed
#### Background
Current UserContractSets continually grow in size. 

Ideally these sets only contain active contracts per user.  Once a contract is completely dead, we can remove it from this set. 

#### AC
1. prune UserContractSets as contracts are completely expired
2. Create tests to confirm this occurs only when a contract is completely dead and no other time.

### PR DESCRIPTION
**Pull request by 0xean**
closes #104

### CODE REVIEW

**Review by johnnyluo — COMMENTED** [review]
If you are concerned about the size of contract set , and clean it up by removing stuff from it , it might not going to work as you expected.  The key value store size will keep growing. 
As @cbarraford mentioned previously on discard , block chain is versioned , append only store , the size will shrink only when `prune` and `compation kicks in, which doesn't happen very often.

what I mean the size will grow?  give you an example ,  assume on block 10 we have a contract set like 
```
[1,2,3,4,5,6,7,8,9,10] 
``` 
and on block 15 , half of the contract ids have been removed  the contract set is 
```
[6,7,8,9,10]
``` 
on block height 15, the key value store will have two contract sets stored , one stored against block height 10, another one store against block height 15.  

when block 10 get pruned , the key value store size will not shrink , because it just been marked as tombstone ,  only when compaction kicks in , then contract sets store against block 10 will be completely removed.  Compaction doesn't happen often , because it is very expensive.

File: x/arkeo/types/keeper.go [hunk 1]
```diff
@@ -89,3 +89,31 @@ func (ct *ContractType) UnmarshalJSON(b []byte) error {
...
+
+	if len(userContractSet.ContractSet.ContractIds) == 0 {
+		return nil, fmt.Errorf("contract set is empty")
+	}
+
+	isFound := false
+	for i, contractId := range userContractSet.ContractSet.ContractIds {
+		if contractId == contractIdToRemove {
+			userContractSet.ContractSet.ContractIds = append(userContractSet.ContractSet.ContractIds[:i], userContractSet.ContractSet.ContractIds[i+1:]...)
+			isFound = true
```
[the remark this answers is not in the document]
Comment (0xean) [reply x/arkeo/types/keeper.go]: yea, thats cleaner. resolved.

File: x/arkeo/keeper/msg_server_open_contract.go [hunk 2]
```diff
@@ -55,6 +55,10 @@ func (k msgServer) OpenContractValidate(ctx cosmos.Context, msg *types.MsgOpenCo
 		return err
 	}
 
+	if provider.LastUpdate == 0 {
```
[the remark this answers is not in the document]
Comment (0xean) [reply x/arkeo/keeper/msg_server_open_contract.go:58]: Currently, these are both populated even if the provider doesn't exist in order to create the key. 

https://github.com/arkeonetwork/arkeo/blob/8667f0ede18495521ecd841942de7f26f8d2658c/x/arkeo/keeper/providers.go#L41

**Comment by 0xean** [comment]
> If you are concerned about the size of contract set , and clean it up by removing stuff from it , it might not going to work as you expected. The key value store size will keep growing. As @cbarraford mentioned previously on discard , block chain is versioned , append only store , the size will shrink only when `prune` and `compation kicks in, which doesn't happen very often.
> 
> what I mean the size will grow? give you an example , assume on block 10 we have a contract set like
> 
> ```
> [1,2,3,4,5,6,7,8,9,10] 
> ```
> 
> and on block 15 , half of the contract ids have been removed the contract set is
> 
> ```
> [6,7,8,9,10]
> ```
> 
> on block height 15, the key value store will have two contract sets stored , one stored against block height 10, another one store against block height 15.
> 
> when block 10 get pruned , the key value store size will not shrink , because it just been marked as tombstone , only when compaction kicks in , then contract sets store against block 10 will be completely removed. Compaction doesn't happen often , because it is very expensive.

This is not a concern about the size of the KV store, but more so the search for an active contract for a specific user.  Currently we iterate through contracts to find a match on an active contract for a given provider and chain.  This is done when opening a new contract to enforce no duplication of contracts. So there is a tradeoff here.... we can leave the set unbounded with all contracts of the users active or not, but pay the penalty when we need to search the set, OR we can utilize this PR and keep the brute force search much more efficient.

**Review by asamere — APPROVED** [review]

**Pull request merged**